### ISSUE TITLE
cuda version of IndRNNCell?

### ISSUE THREAD

**Issue by wwu** [Sunnydreamrain/IndRNN_pytorch#9]
Hi, is there any chance to implement a cuda version of IndRNNCell? the purpose is to speed up processing variable length sequences.

**Comment by Sunnydreamrain** [comment]
Hi,

There is already a CUDA version, cuda_IndRNN_onlyrecurrent.py.   

https://github.com/Sunnydreamrain/IndRNN_pytorch/blob/master/cuda_IndRNN_onlyrecurrent.py

**Comment by wwu** [comment]
Thanks for the reply. I asked for cuda version of IndRNNCell because my input sequence length varies in each batch. I was trying to extend the model with PackSequence input, where cuda_IndRNN_onlyrecurrent doesn't look like a fit.

Pls let me know if I miss anything.

**Comment by Sunnydreamrain** [comment]
Hi,

It works for variable length input as well. I have also used it. In the cuda_IndRNN_onlyrecurrent file, the length is calculated based on the input, not as an input argument. It can also work as a cell to process sequence 1.     
You might need to be careful of the batch normalization modeule though for different lengths.

**Comment by wwu** [comment]
Thanks, I will try out the cuda version as a cell replacement. 
You are right about batch norm, training loss gets steadily lower while val loss fluctuates. Any idea about how to solve this? Right now I am just disabling batch norm which kind of works.

**Comment by wwu** [comment]
did a quick testing on cuda implementation. it's about 5 times faster than non_cuda version in fixed_length input, but slower in variable length input. Below is the pseudo code: 
```
    def forward_pack_sequence(self, p_seq):
        assert isinstance(p_seq, nn.utils.rnn.PackedSequence)

        # split packsequence data with batchsizes
        splits = p_seq.data.split(tuple(p_seq.batch_sizes), dim=0)
        # initialize h0
        first_batch = splits[0]
        h0 = `first_batch.data.new(first_batch.size(-2),first_batch.size(-1)).zero_().contiguous()`
        # run through each timestep and collect hiddens
        outputs = []
        hx_cell = h0
        for ts in splits:
            hx_cell = hx_cell[:len(ts)]
            hx_cell = IndRNN_Compute_GPU.apply(ts, self.weight_hh, hx_cell) #self.indrnn_cell(ts, hx_cell)
            outputs.append(hx_cell)
        # rebuild hidden into PackedSequence
        return PackedSequence(torch.cat(outputs, dim=0), p_seq.batch_sizes, p_seq.sorted_indices, p_seq.unsorted_indices)
```

**Comment by Sunnydreamrain** [comment]
Hi,

Regarding the speed under variable length input, it is probably true. The cuda cell only optimizes the operations over time so that the kenrels donot need to be called everytime. When it is used as a cell, it should be the same as the pytorch version.   

 Regarding the batch norm, it is all right to use it. When the learning rate decays to a small value, the statistics will become stable and the performance may no longer fluctuate. So you only need to set a large patience in the training process to avoid the learning rate dropping too earlier.

**Comment by wwu** [comment]
Thanks. I guess I'll stick to non-cuda version for now.
